### PR TITLE
Yuliiaspryn/homework7

### PR DESCRIPTION
**Pull request by yspryn**

### CODE REVIEW

File: tests/yspryn/test_retrying_decorator_hw7.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+def test_my_function(mocker):
+    mocker.patch.object(random, 'random', return_value=0.6)
+    assert unstable_function() == 0.6
+
+    mocker.patch.object(random, 'random', side_effect=[0.4, 0.3, 0.2, 0.25, 0.6])
+    assert unstable_function() == 0.6
+    assert random.random.call_count == 5
+
+    mocker.patch.object(random, 'random', side_effect=[0.1, 0.15, 0.14, 0.2, 0.33, 0.49, 0.44, 0.33, 0.2, 0.475])
```
Comment (fkudinov-mn) [inline tests/yspryn/test_retrying_decorator_hw7.py:14]: 6 elements here enough

File: yspryn/hw7/caching.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+def decorator_answers(func):
+    remember_result = {}
+
+    def wrapper(n: int):
+        if n not in remember_result:
+            res = func(n)
+            remember_result[n] = res
+        else:
+            return remember_result[n]
+        return remember_result
```
[the remark this answers is not in the document]
Comment (fkudinov-mn) [reply yspryn/hw7/caching.py:16]: why do you need to return   remember_result. - which is dict with all answers

File: yspryn/hw7/sorting.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+start = time.time_ns()
+print(time.time_ns() - start)
+"""
+
+
+from operator import gt, lt, le, ge
+
+
+def time_of_func_execution(func):
+    def wrapper(iterable, reverse: bool = True, key: callable = None):
```
Comment (fkudinov-mn) [inline yspryn/hw7/sorting.py:15]: *args, **kwargs

File: yspryn/hw7/sorting.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        start = time.time_ns()
+        res = func(iterable, reverse, key)
```
Comment (fkudinov-mn) [inline yspryn/hw7/sorting.py:17]: *args, **kwargs

File: tests/yspryn/test_caching_hw7.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import pytest
+from yspryn.hw7 import caching
+
+
+def test_remember_answer_decorator(mocker):
+    spy = mocker.spy(caching, 'fibo')
+    caching.fibo(3)
+    assert spy.call_count == 3
+    assert spy.spy_return == 3
+    caching.fibo(6)
```
Comment (fkudinov-mn) [inline tests/yspryn/test_caching_hw7.py:10]: assert caching.fibo(6) == 13

File: yspryn/hw7/caching.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+def decorator_answers(func):
+    remember_result = {}
+
+    def wrapper(n: int):
+        if n in remember_result:
+            return remember_result[n]
+        res = func(n)
+        remember_result[n] = res
+        return res
```
Comment (fkudinov-mn) [inline yspryn/hw7/caching.py:15]:
```
        if n not  in remember_result:
            remember_result[n] = func(n)
        returnremember_result[n]
```

or

**Review by kudinov-fedor — APPROVED** [review]

**Pull request merged**